[FSDP] assert to runtime error

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #85496
* #85495
* #86343
* #86337
* __->__ #86336
* #86331

Prefer raising an error over `assert` which should mostly to indicate a developer bug, but user can cause this error path.

rohan-varma committed:
[FSDP] assert to runtime error

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] assert to runtime error"


Prefer raising an error over `assert` which should mostly to indicate a developer bug, but user can cause this error path.

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -3601,7 +3601,10 @@ def no_sync(self) -> Generator:
-        assert self._is_root, "`no_sync()` on inner FSDP instances is not supported"
+        if not self._is_root:
+            raise RuntimeError(
+                "`no_sync()` on inner FSDP instances is not supported. Please call `no_sync()` on root FSDP module."
```
nit: In case you are okay with re-pushing:
```suggestion
                "`no_sync()` on inner FSDP instances is not supported. Please call `no_sync()` on the root FSDP module."
```

awgu reviewed (approved):
LGTM!

rohan-varma committed:
Update on "[FSDP] assert to runtime error"


Prefer raising an error over `assert` which should mostly to indicate a developer bug, but user can cause this error path.

[ghstack-poisoned]
